### PR TITLE
implement Folder datastore

### PR DESCRIPTION
**Pull request by Gustry**
First implementation for the Folder datastore.

Ready to merge this first PR.
I'm going to add some child class to manage  PostGIS/GeoPackage.

We can write shapes and tiffs into a folder and get the list of layers in the datastore.

### CODE REVIEW

File: safe/datastore/datastore.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+# coding=utf-8
+"""
+InaSAFE Disaster risk assessment tool by AusAid - **Clipper test suite.**
```
Comment (ismailsunni) [inline safe/datastore/datastore.py]: I knew it's from copy and paste :), we should give a better docstring.

File: safe/datastore/folder.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
+# coding=utf-8
+"""
+InaSAFE Disaster risk assessment tool by AusAid - **Clipper test suite.**
```
Comment (ismailsunni) [inline safe/datastore/folder.py]: More suitable docstring

File: safe/datastore/test/test_folder.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    def setUp(self):
+        pass
+
+    def tearDown(self):
+        pass
+
+    def test_folder_datastore(self):
+        """Test if we can store shapefiles."""
+        path = QDir(mkdtemp())
+        ds = Folder(path)
```
Comment (ismailsunni) [inline safe/datastore/test/test_folder.py]: perhaps it should be `data_storage` to make it clearer.

Comment (Gustry) [inline safe/datastore/test/test_folder.py]: You mean to call it 'FolderDataStore' ?

Comment (ismailsunni) [inline safe/datastore/test/test_folder.py]: My original point is the `ds` (the variable name) should be something clearer, like `data_store`.

File: safe/datastore/datastore.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    def __init__(self, uri):
+        """
+        Constructor for the DataStore.
+
+        The datastore can be used in three different ways :
+        - a PostGIS connection
+        - a folder to store a shapefile
+        - a geopackage file.
+
```
Comment (timlinux) [inline safe/datastore/datastore.py]: in future we may add support for others...

File: safe/datastore/folder.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+from safe.datastore.datastore import DataStore
+from safe.common.exceptions import ErrorDataStore, ShapefileCreationError
+
+EXTENSIONS = ('shp', 'asc', 'tiff', 'tif')
+
+
+class Folder(DataStore):
+    """
+    Folder DataStore
+
```
Comment (timlinux) [inline safe/datastore/folder.py]: A folder based data store is a collection of shape files and tiff images stored in a common folder.

File: safe/datastore/folder.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        """Check if the folder is writable.
+
+        :return: If it's writable or not.
+        :rtype: bool
+
+        .. versionadded:: 4.0
+        """
+        return QFileInfo(self._uri.absolutePath()).isWritable()
+
+    def is_raster_supported(self):
```
Comment (timlinux) [inline safe/datastore/folder.py]: supports_rasters

File: safe/datastore/geopackage.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+        :return: If it's writable or not.
+        :rtype: bool
+
+        .. versionadded:: 4.0
+        """
+        # Fixme, need to check DB permissions ?
+        return self._uri.absolutePath().isWritable()
+
+    def is_raster_supported(self):
```
Comment (timlinux) [inline safe/datastore/geopackage.py]: supports_raster

File: safe/datastore/geopackage.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        :rtype: (bool, str)
+
+        .. versionadded:: 4.0
+        """
+
+        # Fixme
+        # if not self.is_writable():
+        #    return False, 'The destination is not writable.'
+
+        geom = QGIS_OGR_GEOMETRY_MAP[vector_layer.wkbType()]
```
Comment (timlinux) [inline safe/datastore/geopackage.py]: geometry

File: safe/datastore/geopackage.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+        srs = osr.SpatialReference()
```
Comment (timlinux) [inline safe/datastore/geopackage.py]: spatial_reference

File: safe/datastore/geopackage.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        qgis_crs = vector_layer.crs().authid()
```
Comment (timlinux) [inline safe/datastore/geopackage.py]: coordinate_reference

File: safe/datastore/geopackage.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        srs.ImportFromEPSG(int(qgis_crs.split(':')[1]))
+
+        self.datasource.CreateLayer(layer_name, srs, geom)
+        uri = u'{}|layerid=0'.format(self.uri.absoluteFilePath())
+        vl = QgsVectorLayer(uri, layer_name, u'ogr')
```
Comment (timlinux) [inline safe/datastore/geopackage.py]: vector_layer

**Pull request merged**